helix-core/src/main/java/org/apache/helix/HelixProperty.java

```diff
@@ -39,7 +39,7 @@
  * A wrapper class for ZNRecord. Used as a base class for IdealState, CurrentState, etc.
  */
 public class HelixProperty {
```

Comment:
Correct me if I'm wrong, but because we "deep copy" `HelixProperty` and its internal `ZNRecord`, throughout Helix we can treat this as such, and that is, we can assume that it's immutable. But in the case that we do not clone it and simply take after the internal `ZNRecord` reference, such assumptions no longer hold and therefore may open up avenues where issues could creep in. And this particular change doesn't allow us to know whether its internal `ZNRecord` reference is being shared across different `HelixProperty` instances or not.

Could we think of alternative ways to achieve this if you think this is necessary? For example, add a subclass of `HelixProperty` that shallow copies `ZNRecord`, so that there is no ambiguity?

Reply:
Of course, this does not clone the `ZNRecord`, so code which assumes the `ZNRecord` has been mutated will not work properly, but these constructors are made available for advanced users of Helix, like Apache Pinot. In our code, the copy is needless and shows up as a surprisingly noticeable cost, which I would like to avoid incurring. This is why I took care to document the constructors; use at your own risk, take responsibility for it, but make it possible not to incur the cost.

Subclassing would not work because `HelixProperty` would need to be constructed to construct the subclass, which would result in a clone. It could work the other way around though, but this flag feels like a simple, easy to document, and backward compatible solution.

Reply:
Yes, I understand the motivation behind this. That still doesn't give us a way to tell if the object is thoroughly immutable (because the underlying ZNRecord has been deep-copied) - I was thinking maybe adding a final `boolean` that can tell us whether it's mutable or immutable? Do you think that would be useful for the users of `HelixProperty`?

Reply:
But you don't have that guarantee anyway - `HelixProperty` exposes a reference to its `_record` with `getRecord` and the `ZNRecord` itself is mutable.

Reply:
In that case, I think this change is okay since we didn't have that guarantee to begin with. Thanks!